c/src/atom.rs

```diff
@@ -69,6 +72,101 @@ pub extern "C" fn exec_error_free(error: *mut exec_error_t) {
     unsafe{ drop(Box::from_raw(error)); }
 }
 
+// bindings
+#[no_mangle]
+pub unsafe extern "C" fn bindings_new() -> *mut bindings_t {
+    bindings_into_ptr(Bindings::new())
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn bindings_free(bindings: *mut bindings_t) {
+    // drop() does nothing actually, but it is used here for clarity
+    drop(Box::from_raw(bindings));
+}
+
+#[no_mangle]
+pub extern "C" fn bindings_clone(bindings: *const bindings_t) -> *mut bindings_t {
+    bindings_into_ptr(unsafe{ &(*bindings) }.bindings.clone())
+}
+
+#[no_mangle]
+pub extern "C" fn bindings_to_str(bindings: *const bindings_t, callback: c_str_callback_t, context: *mut c_void) {
+    let bindings = unsafe{ &(*bindings).bindings };
+    let s = str_as_cstr(bindings.to_string().as_str());
+    callback(s.as_ptr(), context);
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn bindings_eq(bindingsa: *const bindings_t, bindingsb: *const bindings_t) -> bool {
+    (*bindingsa).bindings == (*bindingsb).bindings
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn bindings_traverse(bindings: * const bindings_t, callback: lambda_t<* const var_atom_t>, context: *mut c_void) {
+    (*bindings).bindings.iter().for_each(|(var, atom)|  {
+            let name = string_as_cstr(var.name());
+            let var_atom = var_atom_t {
+                var: name.as_ptr(),
+                atom: atom_into_ptr(atom)
+            };
+            callback(&var_atom, context);
+        }
+    )
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn bindings_add_var_binding(bindings: * mut bindings_t, atom: *const var_atom_t) -> bool {
+    let bindings = &mut(*bindings).bindings;
+    let var = VariableAtom::new(cstr_into_string ((*atom).var));
+    let atom = (*(*atom).atom).atom.clone();
```

Comment:
Here we should be able to remove `clone` and use `ptr_into_atom` instead:
```rs
    let atom =ptr_into_atom(atom);
```
Atom should be already cloned before calling `bindings_add_var_binding` in `hyperonpy.cpp`.